### ISSUE TITLE
Add equivalent to TeX glue stretching (dynamic vertical spacing)

### ISSUE THREAD

**Issue by GamerBene19** [typst/typst#3442]
### Description

In TeX there is a concept called glue stretching (see [this StackOverflow answer](https://tex.stackexchange.com/a/64759) for more details).
As far as I understand it, the basic idea is that instead of having a fixed height vspace, one is able to provide a set size and (optional) plus and/or minus modifiers to specify a range of "acceptable" vspace from which the compiler is able to choose what it sees fit.

### Use Case

Honestly I don't know what exactly the usecase for this is (personally I'd prefer consistent spacing throughout the document (or parts of it). However, I think there could be certain cases where this would help with lay outing the document.
I'm mainly just creating this issue to throw it out there for discussion.

**Comment by davidwoodburn** [comment]
I would strongly recommend against this. Vertical justification (as it is called) is a bad idea and directly opposed by document layout books. It is one of my biggest frustrations with LaTeX. For the weird case where something like this is desired, I would suggest instead that manually inserted, flexible leading be used.

**Comment by JustForFun88** [comment]
I am also against this idea. It looks very strange when the baseline is different on different pages of the book, especially on the spreads. I don't remember Adobe InDesign or Scribus doing anything like this. On the contrary, InDesign and Scribus maintain the same baseline on all pages.

LaTex is certainly not bad at formatting formulas, but nothing more.

**Comment by Munchotaur** [comment]
I've done this in InDesign ("vertical justification"), also in the old days of Ventura Publisher and some other layout software. You certainly have to handle it right or you'll get vile-looking gaps, but it's handy for things like multi-column glossaries with short entries in small type, or small ads, or ...

InDesign maintains common baseline grids *if you instruct it to*. Just like vertical justification, it isn't enforced by default and nor should it be. For more free-form page layouts, grids needn't be the whole of the law either.

Good or bad (and in general I do much prefer grids), purist or not, things like vertical justification have their place in real-world typesetting. To simply wave them away as bad practice is to make a heck of a lot of assumptions about other people's legitimate use cases.

**Comment by davidwoodburn** [comment]
To your very mature (and probably correct) comments I say, “Well, they should all do it my way if they know what’s good for them.” In all seriousness, though, I would just like to make sure it’s not a default I have to fight with, like in LaTeX.

**Comment by Munchotaur** [comment]
> ... I would just like to make sure it’s not a default I have to fight with, like in LaTeX.

Absolutely. Defaults should be kept to a bare and sensible minimum (whatever that is!) and should be easy to override. For the most part, I think Typst does a pretty good job of that. I don't know if there's a way of setting more permanent per-user defaults via a config file, but that would be nice to have.

**Comment by emilyyyylime** [comment]
Typst aims to keep documents as reproducible as possible, so any implicitly included config files would go against that philosophy. You can however just include a local package in every project you make

**Comment by memeplex** [comment]
I arrived here with this use case in mind: slideshows. I usually have some bullets in a slide column that I like to be centered (horizon) and stretch to use the available vertical space (up to a sensible limit, that's why they are also centered). I could compute the available space and insert fixed spacing between the items, but I believe it's a use case for this request too.

**Comment by laurmaedje** [comment]
This sounds like something that could be done with vertical fr spacing.

**Comment by memeplex** [comment]
I believe you can get close but not exactly there. AFAICR with glue you can control how far the stretching (or shrinking) may go. In my example, I wouldn't like if the items in the list became too spaced. I'm not saying mine is a highly compelling use case, but I think it's a pretty natural one.

From the TeXBook (there is a nice picture and further explanations, but I'm not sure about the copyright so I'm stopping here):

> Glue has three attributes, namely its natural space, its ability to stretch, and its ability to shrink.

**Comment by eltos** [comment]
Another use case of this is **alignment of last line in two column layouts**. Figures, headings or subsection headings cause the lines to drift apart anyway, and orphan and widow prevention can cause large blank spaces in either of the columns.

Note that this is somewhat equivalent to a very precise form of column balancing on a fully filled page (compare #466 for balancing on partly filled pages).

<details>
<summary>Show code</summary>

```typst
#set page(paper: "a6", flipped: true, columns: 2, margin: 2em)
#set par(justify: true)

#place(top, rect(width: 100%, height: 100%, stroke: (paint: red, thickness: 1pt)))

= Section
#lorem(30)

== Subsection
#lorem(20)

== Subsection
#lorem(20)

#place(top, rect(width: 100%, height: 100%, stroke: (paint: red, thickness: 1pt)))

= Section
#lorem(40)

== Subsection
#lorem(20)
```

</details>


|Fixed spacings (default)|Dynamic stretching|
|-|-|
|![column balance-1](https://github.com/user-attachments/assets/38c95272-45cc-4c44-be76-1bfeb9d02fba)|![column balance(1)](https://github.com/user-attachments/assets/8f09667a-e12e-4da3-b345-20a9cb759682)

**Comment by TheZoq2** [comment]
Until this is supported natively, I figured out a solution that looks OK for my use case.

Here is what it looks like

![image](https://github.com/user-attachments/assets/06737dc3-5e0e-41c3-bb3f-0f67939feec9)


The trick is to add 3 blocks around the heading, one on each side with fixed height for the minimum spacing, then one on top with fractional spacing

```
  show heading: it => {
    context {
      // Find out the final number of the heading counter.
      let levels = counter(heading).at(here())
      let no-numbering = it.body in ([Abstract], [Acknowledgments], [Acknowledgment])
      let deepest = if levels != () { levels.last() } else { 1 }

      // Allow headings to stretch their space before themselves to make columns end at the
      // end of the page
      if not it.body in ([Abstract],) {
        block(spacing: 0%)
        block(width: 100%, spacing: 0%, height: 1fr)
        block(width: 100%, spacing: 0%, height: 6pt)
      }

      let heading_block(size) = {
        block(text(size: size, {
          if it.numbering != none and not no-numbering{ 
            numbering(it.numbering, ..levels)
            h(spacing, weak: true)
          }
          it.body
        }), spacing: 0%)
      }
      if it.level == 1 {
        heading_block(12pt)
      } else if it.level == 2 {
        heading_block(10pt)
      } else {
        heading_block(10pt)
      }
    }

    // Ensure 6pt spacing between headings and their content 
    block(width: 100%, height: 6pt, spacing: 0%)
  }
```

![image](https://github.com/user-attachments/assets/0ed98817-fd3e-401f-a65f-8bdf572e5b98)

**Comment by Munchotaur** [comment]
That's a measured (!) approach. I'll have to give it a try but it looks sensible.
